File: 02_ImageFiltering/02_LinearFiltering/03_GaussianFilter/gaussian2dpy/gaussianPyramid.py
import cv2
import numpy as np

def apply_gaussian_filter(image, kernel_size, sigma):
    return cv2.GaussianBlur(image, (kernel_size, kernel_size), sigma)

def downsample_image(image):
    return cv2.resize(image, (image.shape[1] // 2, image.shape[0] // 2))

def create_gaussian_pyramid(image, levels):
    pyramid = [image]
    for _ in range(1, levels):
        blurred = apply_gaussian_filter(pyramid[-1], 5, 1.0)
        downsampled = downsample_image(blurred)
        pyramid.append(downsampled)
    return pyramid

if __name__ == "__main__":
    image = cv2.imread("../01_Images/original_image.png", cv2.IMREAD_GRAYSCALE)
    if image is None:
        print("Error: Image not found!")
        exit(1)
    
    gaussian_pyramid = create_gaussian_pyramid(image, 3)
    
    cv2.imwrite("../01_Images/gaussian_pyramid_level1.png", gaussian_pyramid[1])
    cv2.imwrite("../01_Images/gaussian_pyramid_level2.png", gaussian_pyramid[2])
    
    print("Gaussian pyramid generated successfully!")

File: 02_ImageFiltering/02_LinearFiltering/03_GaussianFilter/gaussianPyramid.cpp
#include <opencv2/opencv.hpp>
#include <iostream>

cv::Mat applyGaussianFilter(const cv::Mat& image, int kernelSize, double sigma) {
    cv::Mat result;
    cv::GaussianBlur(image, result, cv::Size(kernelSize, kernelSize), sigma);
    return result;
}

cv::Mat downsampleImage(const cv::Mat& image) {
    cv::Mat downsampled;
    cv::resize(image, downsampled, cv::Size(image.cols / 2, image.rows / 2));
    return downsampled;
}

std::vector<cv::Mat> createGaussianPyramid(const cv::Mat& image, int levels) {
    std::vector<cv::Mat> pyramid;
    pyramid.push_back(image);
    for (int i = 1; i < levels; ++i) {
        cv::Mat blurred = applyGaussianFilter(pyramid.back(), 5, 1.0);
        cv::Mat downsampled = downsampleImage(blurred);
        pyramid.push_back(downsampled);
    }
    return pyramid;
}

int main() {
    cv::Mat image = cv::imread("../../01_Images/Checker_original.png", cv::IMREAD_GRAYSCALE);
    if (image.empty()) {
        std::cerr << "Error: Image not found!" << std::endl;
        return -1;
    }
    
    std::vector<cv::Mat> gaussianPyramid = createGaussianPyramid(image, 3);
    
    cv::imwrite("../../01_Images/gaussian_pyramid_level1.png", gaussianPyramid[1]);
    cv::imwrite("../../01_Images/gaussian_pyramid_level2.png", gaussianPyramid[2]);
    
    return 0;
}
